output/batch/stdout.go

```diff
@@ -0,0 +1,171 @@
+package batch
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"sort"
+	"time"
+
+	tm "github.com/buger/goterm"
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/storage"
+)
+
+type TableFormatter func(w io.Writer, records []*execution.Record, watermark time.Time, err error) error
+
+type RecordsLister interface {
+	ListRecords(ctx context.Context, tx storage.StateTransaction) ([]*execution.Record, error)
+	GetWatermark(ctx context.Context, tx storage.StateTransaction) (time.Time, error)
+	GetEndOfStream(ctx context.Context, tx storage.StateTransaction) (bool, error)
+	GetErrorMessage(ctx context.Context, tx storage.StateTransaction) (string, error)
+}
+
+type LiveTablePrinter struct {
+	recordsLister  RecordsLister
+	stateStorage   storage.Storage
+	tableFormatter TableFormatter
+}
+
+func NewLiveTablePrinter(stateStorage storage.Storage, recordsLister RecordsLister, tableFormatter TableFormatter) *LiveTablePrinter {
+	return &LiveTablePrinter{
+		recordsLister:  recordsLister,
+		stateStorage:   stateStorage,
+		tableFormatter: tableFormatter,
+	}
+}
+
+func (printer *LiveTablePrinter) Run(ctx context.Context) error {
+	for range time.Tick(time.Second / 4) {
+		tx := printer.stateStorage.BeginTransaction()
+
+		records, err := printer.recordsLister.ListRecords(ctx, tx)
+		if err != nil {
+			return errors.Wrap(err, "couldn't list records")
+		}
+
+		if len(records) > 0 && !records[0].EventTimeField().Empty() {
+			sort.SliceStable(records, func(i, j int) bool {
+				return records[i].EventTime().AsTime().Before(records[j].EventTime().AsTime())
```

Comment:
Why do you assume that if the first record has an `EventTimeField` then all records do?

Reply:
Because the EventTime is something known for a stream at physical plan time. So we know it holds. If one record in a stream has an event time, then the stream is event timed, then all have it.

I didn't pass the event time in the constructor because this is just minor beautifying of the output, and I may want to rewrite it in the future to be more akin to the stream output (keep the input ordering).

Reply:
OK